File: Data Structures & Algorithms/validate-parentheses/submission-1.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0: return False 
        stack = []
        
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
            elif len(stack) == 0:
                return False
            else: 
                if (c == ')' and stack[-1] == '(') or (c == '}' and stack[-1] == '{') or (c == ']' and stack[-1] == '[') :
                    stack.pop()
                else:
                    return False

        return len(stack) == 0

File: Data Structures & Algorithms/valid-tree/submission-1.cpp
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        if (edges.size() > n-1) return false;
        unordered_map<int, unordered_set<int>> graph;
        for (vector<int> &v: edges)
        {
            graph[v[0]].insert(v[1]);
            graph[v[1]].insert(v[0]);
        }

        unordered_set<int> visited;
        queue<pair<int, int>> q;
        q.push({0, -1});
        visited.insert(0);

        while (!q.empty())
        {
            auto &[node, parent] = q.front();
            q.pop();

            for (int neighbour: graph[node]){
                if (neighbour == parent)
                {
                    continue;
                }

                if (visited.count(neighbour) != 0)
                {
                    return false;
                }

                visited.insert(neighbour);
                q.push({neighbour, node});
            }
        }

        return visited.size() == n;
    }
};
